pkg/logutil/pluggable_logger.go

```diff
@@ -0,0 +1,143 @@
+func (l *PluggableLogger) PlugLogger(logger *zap.Logger) {
+	atomic.StorePointer(&l.logger, unsafe.Pointer(logger))
+}
+
+// UnplugLogger unplugs the logger. If it has been replaced, nothing will be done.
+func (l *PluggableLogger) UnplugLogger(logger *zap.Logger) {
+	atomic.CompareAndSwapPointer(&l.logger, unsafe.Pointer(logger), nil)
+}
+
+// Debug logs a message at DebugLevel. The message includes any fields passed
+// at the log site, as well as any fields accumulated on the logger.
+func (l *PluggableLogger) Debug(msg string, fields ...zap.Field) {
+	if logger := l.GetLogger(); logger != nil {
+		logger.Debug(msg, fields...)
+	}
+}
+
+// Info logs a message at InfoLevel. The message includes any fields passed
+// at the log site, as well as any fields accumulated on the logger.
+func (l *PluggableLogger) Info(msg string, fields ...zap.Field) {
+	if logger := l.GetLogger(); logger != nil {
+		logger.Info(msg, fields...)
+	}
+}
+
+// Warn logs a message at WarnLevel. The message includes any fields passed
+// at the log site, as well as any fields accumulated on the logger.
+func (l *PluggableLogger) Warn(msg string, fields ...zap.Field) {
+	if logger := l.GetLogger(); logger != nil {
+		logger.Warn(msg, fields...)
+	}
+}
+
+// Error logs a message at ErrorLevel. The message includes any fields passed
+// at the log site, as well as any fields accumulated on the logger.
+func (l *PluggableLogger) Error(msg string, fields ...zap.Field) {
+	if logger := l.GetLogger(); logger != nil {
+		logger.Error(msg, fields...)
+	}
+}
+
+// Panic logs a message at PanicLevel. The message includes any fields passed
+// at the log site, as well as any fields accumulated on the logger.
+//
+// The logger then panics, even if logging at PanicLevel is disabled.
+func (l *PluggableLogger) Panic(msg string, fields ...zap.Field) {
+	if logger := l.GetLogger(); logger != nil {
+		logger.Panic(msg, fields...)
+	}
+}
+
+// Fatal logs a message at FatalLevel. The message includes any fields passed
+// at the log site, as well as any fields accumulated on the logger.
+//
+// The logger then calls os.Exit(1), even if logging at FatalLevel is
+// disabled.
+func (l *PluggableLogger) Fatal(msg string, fields ...zap.Field) {
+	if logger := l.GetLogger(); logger != nil {
+		logger.Fatal(msg, fields...)
+	}
+}
+
+type loggerManager struct {
+	sync.Mutex
+	Loggers map[string]*PluggableLogger
+}
+
+func (m *loggerManager) Init() {
+	m.Lock()
+	defer m.Unlock()
+	m.Loggers = make(map[string]*PluggableLogger)
+}
+
+// GetPluggableLogger gets a pluggable zap.Logger.
+func (m *loggerManager) GetPluggableLogger(name string, createIfNotExist bool) *PluggableLogger {
+	m.Lock()
+	defer m.Unlock()
+	l, ok := m.Loggers[name]
+	if !ok {
+		if !createIfNotExist {
+			return nil
+		}
+		l = &PluggableLogger{name: name}
+		m.Loggers[name] = l
+	}
+	return l
+}
+
+var globalLoggerManager loggerManager
+
+func init() {
```

Comment:
Is there anyway to avoid it?

Reply:
In the example, `GetPluggableLogger` will be used to generate some new global loggers.